##### CSE 5243 - Introduction to Data Mining
## Homework 5: Association Analysis
- Semester: Fall 2022
- Instructor: Greg Ryslik
- Section: Wednesday 12:45 PM
- Student Name: John Smith
- Student.#: smith.2
***

# Introduction

### Objectives

In this lab, you will use the "hw5_data.csv" dataset provided on Carmen to find potential association rules.


The objectives of this assignment are:
1.    Practice the Association Analysis content we covered this semester.
2.    Understand “why” the particular topics, techniques, etc., are important from a practical perspective.
3.    Understand how to choose and use appropriate tools to solve the provided problems.

### The Dataset
- This workbook contains is a market basket dataset containing 50 transactions, drawing from a universe of six items: Apples, Bananas, Carrots, Donuts, Eggs, Fish.  For simplicity, use the short form “A, B, C, D, E, F” for the items.
- There is one csv file that captures the data in "long format". Specifically, every row corresponds to the transaction id and the item. If the specific transaction id has multiple items, you will have multiple rows in your data.
- You can use it however you like but it is recommended you convert into the one-hot-encoded datastructure we used in class. This will allow you to easily use the mlxtend package.

### Proper answers
- To make everyone's lives a little easier, when writing itemsets and rules, please list them in lexagraphical order:
  {A}, {B}, {A,B}, {A,C}, {A,B,C},…
  {A,B,C}->{D}, {A,B}->{C,D}

### Collaboration
For this assignment, you should work as an individual. You may informally discuss ideas with classmates, but your work should be your own.

### What you need to turn in:
1)	Turn in this Jupyter Notebook BOTH in notebook format and in pdf format. Failure to submit one or the other will result in points being deducted.
  - Submit your hw as HW5_Surname_DotNumber.zip
  
2)  Feel free to use the **mlxtend** package to help to help enumerate all possible combinations.

3)  If the question asks you to compute all possible rules, back up the calculation with a "formula" approach as well (similar to how we did in class/slides). This will act as "showing your work".
***

***
# Section 1: Getting Ready (20%)
1A) Load the data, and get it ready for association analysis. Do this with convenient python helper methods as appropriate. Feel free to use the tools we learned in class. HINT: If you're code looks like you're writing it in C++/Java/etc with lots of messy for loops, step back and re-evaluate. 
    
    - Make the data one-hot encoded.
***

In [20]:
#pip install mlxtend
import numpy as np
import pandas as pd
import mlxtend as mlx

df = pd.read_csv("hw5_data.csv")
df

C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


TxId   ItemId
0       1   Donuts
1       1     Eggs
2       2  Bananas
3       2  Carrots
4       2     Fish
..    ...      ...
126    50  Bananas
127    50  Carrots
128    50   Donuts
129    50     Eggs
130    50     Fish

[131 rows x 2 columns]

In [21]:
## Number of unique items:
items = set()
for col in df[['ItemId']]:
    items.update(df[col].unique())

print(items)

{'Carrots', 'Bananas', 'Apples', 'Fish', 'Donuts', 'Eggs'}


C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [22]:
## Number of unique ids:
ids = set()
for col in df[['TxId']]:
    ids.update(df[col].unique())

print(ids)

{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50}


C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [23]:
from mlxtend.preprocessing import TransactionEncoder

te = mlx.preprocessing.TransactionEncoder()

a=[]
for i in ids:
    a.append(list(df[df['TxId']==i]['ItemId']))
te_array = te.fit(a).transform(a)
ohe_df= pd.DataFrame(te_array, columns=te.columns_)
ohe_df

C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Apples  Bananas  Carrots  Donuts   Eggs   Fish
0    False    False    False    True   True  False
1    False     True     True   False  False   True
2    False     True    False   False   True  False
3     True     True     True   False  False   True
4     True     True    False    True  False  False
5    False    False     True   False   True  False
6    False    False    False    True  False  False
7     True    False    False    True  False   True
8    False    False     True   False  False   True
9     True     True     True   False  False   True
10   False     True    False   False   True  False
11   False     True     True   False  False  False
12   False    False    False    True  False  False
13   False     True    False   False  False  False
14   False    False    False    True   True  False
15   False    False    False    True   True  False
16   False     True     True   False  False   True
17   False     True     True   False   True  False
18    True     True     True   False  False   True
19    True     True    False    True  False  False
20   False    False     True   False   True   True
21   False    False    False    True  False  False
22    True    False     True    True  False   True
23   False    False     True   False  False   True
24    True     True     True   False  False   True
25   False     True    False   False   True   True
26   False     True     True   False   True  False
27   False    False    False    True  False  False
28    True     True    False   False  False  False
29   False    False    False    True   True  False
30   False     True    False    True   True  False
31   False     True     True   False  False   True
32   False     True    False   False   True  False
33    True     True     True   False  False   True
34    True     True    False    True  False  False
35   False    False     True   False   True   True
36   False    False    False    True  False  False
37    True    False    False    True  False   True
38   False    False     True   False  False   True
39    True     True     True   False  False   True
40    True     True    False   False   True  False
41   False     True     True   False  False  False
42    True    False    False    True  False  False
43   False     True    False   False  False  False
44   False    False    False    True   True  False
45   False    False    False    True   True  False
46   False     True     True   False  False   True
47   False     True     True   False   True  False
48    True     True     True   False   True   True
49    True     True     True    True   True   True

***
# Section 2: Basic Stats (20%)
2A) Calculate the total number of Itemsets that could be created from a universe of six items. Show your work.

2B) Calculate the total number of rules that can be created from a universe of six items. Show your work.

2C) Calculate the total number of ItemSets that could be created from a universe of 12 items. Show your work.

2D) Calculate the total number of rules that can be created from a universe of 12 items. Show your work.

2E) What do the calculations in 2A-2D tell you / hint at as a potential cause of concern? Hint: Complexity. 
***

### 2A) d=6, total number of Itemsets= 2^6=64

In [24]:
2**6

C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


64

### 2B)d=6, total number of rules =3^6-2^(6+1)+1=602

In [25]:
3**6-2**(6+1)+1

C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


602

### 2C) d=12, total number of Itemsets= 2^12=4096

In [26]:
2**12

C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


4096

### 2D)d=12, total number of rules =3^12-2^(12+1)+1=531441-8192+1=523250

In [27]:
3**12-2**(12+1)+1

C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


523250

### 2E) Hint : When the number of items increase, the total number of itemsets and rules increase exponentialy. The calculate complexity also increases exponentialy.

***
# Section 3: Itemset Generation (20%)
3A) Calculate the number Itemsets that are possible in the data file. 

3B) Calculate the number of Itemsets with MinSup = 2.

3C) Calculate the number of Itemsets with MinSup = 3.

3D) Calculate the number of Itemsets with MinSup = 4. 

3E) Using MinSup = 3, use the Apriori algorithm (feel free to use the package) to list all ItemSets with MinSup = 3. 

3F) Using MinSup = 3, find all the Maximal Frequent Itemsets.

3G) Find all the Closed Itemsets.

3H) Using Minsup = 3, find the Closed Frequent Itemsets. 

3I) How do the Closed Frequent Itemsets compare to the Maximal Frequent Itemsets? 

3J) When one might use the Closed Frequent vs the Maximal Frequent Itemsets?
***

### 3A) Calculate the number Itemsets that are possible in the data file: d=6, total number of Itemsets= 2^6=64

In [28]:
import itertools
list_combinations = list()
for n in range(len(items) + 1):
    list_combinations += list(itertools.combinations(items, n))
len(list_combinations)

C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


64

### 3B) Calculate the number of Itemsets with MinSup = 2.

In [29]:
from mlxtend.frequent_patterns import apriori

f2 = mlx.frequent_patterns.apriori(ohe_df, min_support=0.04, use_colnames=True)
len(f2)

C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


43

### 3C) Calculate the number of Itemsets with MinSup = 3.

In [30]:
f3 = mlx.frequent_patterns.apriori(ohe_df, min_support=0.06, use_colnames=True)
len(f3)

C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


31

### 3D) Calculate the number of Itemsets with MinSup = 4.

In [31]:
f4 = mlx.frequent_patterns.apriori(ohe_df, min_support=0.08, use_colnames=True)
len(f4)

C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


28

### 3E)List all ItemSets with MinSup = 3.

In [32]:
frequent_itemsets_ap = mlx.frequent_patterns.apriori(ohe_df, min_support=0.06, use_colnames=True)
frequent_itemsets_ap

C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support                          itemsets
0      0.34                          (Apples)
1      0.58                         (Bananas)
2      0.48                         (Carrots)
3      0.40                          (Donuts)
4      0.40                            (Eggs)
5      0.42                            (Fish)
6      0.26                 (Apples, Bananas)
7      0.18                 (Carrots, Apples)
8      0.16                  (Apples, Donuts)
9      0.06                    (Apples, Eggs)
10     0.22                    (Apples, Fish)
11     0.34                (Carrots, Bananas)
12     0.10                 (Bananas, Donuts)
13     0.22                   (Bananas, Eggs)
14     0.26                   (Bananas, Fish)
15     0.16                   (Carrots, Eggs)
16     0.36                   (Carrots, Fish)
17     0.16                    (Donuts, Eggs)
18     0.08                    (Fish, Donuts)
19     0.10                      (Fish, Eggs)
20     0.16        (Carrots, Apples, Bananas)
21     0.08         (Apples, Donuts, Bananas)
22     0.06           (Apples, Eggs, Bananas)
23     0.16           (Apples, Fish, Bananas)
24     0.18           (Carrots, Apples, Fish)
25     0.08            (Apples, Fish, Donuts)
26     0.10          (Carrots, Bananas, Eggs)
27     0.24          (Carrots, Bananas, Fish)
28     0.06             (Bananas, Fish, Eggs)
29     0.08             (Carrots, Fish, Eggs)
30     0.16  (Carrots, Apples, Fish, Bananas)

### 3F) Using MinSup = 3, find all the Maximal Frequent Itemsets.

In [33]:
frequent_itemsets_fp=mlx.frequent_patterns.fpmax(ohe_df, min_support=0.06, use_colnames=True)
frequent_itemsets_fp

C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support                          itemsets
0     0.06           (Apples, Eggs, Bananas)
1     0.08         (Apples, Donuts, Bananas)
2     0.08            (Apples, Fish, Donuts)
3     0.16  (Carrots, Apples, Fish, Bananas)
4     0.16                    (Donuts, Eggs)
5     0.06             (Bananas, Fish, Eggs)
6     0.08             (Carrots, Fish, Eggs)
7     0.10          (Carrots, Bananas, Eggs)

### 3G) Find all the Closed Itemsets.

In [34]:
#https://towardsdatascience.com/how-to-find-closed-and-maximal-frequent-itemsets-from-fp-growth-861a1ef13e21
frequent = mlx.frequent_patterns.fpgrowth(ohe_df, min_support=0.001, use_colnames=True)
su = frequent.support.unique()#all unique support count
fredic = {}
for i in range(len(su)):
    inset = list(frequent.loc[frequent.support ==su[i]]['itemsets'])
    fredic[su[i]] = inset
cl = []
for index, row in frequent.iterrows():
    isclose = True
    cli = row['itemsets']
    cls = row['support']
    checkset = fredic[cls]
    for i in checkset:
        if (cli!=i):
            if(frozenset.issubset(cli,i)):
                isclose = False
                break
    
    if(isclose):
        cl.append(row['itemsets'])
cl

C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


[frozenset({'Eggs'}),
 frozenset({'Donuts'}),
 frozenset({'Bananas'}),
 frozenset({'Carrots'}),
 frozenset({'Fish'}),
 frozenset({'Apples'}),
 frozenset({'Bananas', 'Eggs'}),
 frozenset({'Carrots', 'Eggs'}),
 frozenset({'Eggs', 'Fish'}),
 frozenset({'Bananas', 'Carrots', 'Eggs'}),
 frozenset({'Carrots', 'Eggs', 'Fish'}),
 frozenset({'Bananas', 'Eggs', 'Fish'}),
 frozenset({'Donuts', 'Eggs'}),
 frozenset({'Bananas', 'Donuts'}),
 frozenset({'Bananas', 'Donuts', 'Eggs'}),
 frozenset({'Bananas', 'Carrots'}),
 frozenset({'Carrots', 'Fish'}),
 frozenset({'Bananas', 'Fish'}),
 frozenset({'Bananas', 'Carrots', 'Fish'}),
 frozenset({'Apples', 'Bananas'}),
 frozenset({'Apples', 'Fish'}),
 frozenset({'Apples', 'Donuts'}),
 frozenset({'Apples', 'Carrots', 'Fish'}),
 frozenset({'Apples', 'Bananas', 'Carrots', 'Fish'}),
 frozenset({'Apples', 'Bananas', 'Donuts'}),
 frozenset({'Apples', 'Donuts', 'Fish'}),
 frozenset({'Apples', 'Carrots', 'Donuts', 'Fish'}),
 frozenset({'Apples', 'Bananas', 'Eggs'}),

### 3H) Using Minsup = 3, find the Closed Frequent Itemsets.

In [35]:
frequent = mlx.frequent_patterns.fpgrowth(ohe_df, min_support=0.06, use_colnames=True)
su = frequent.support.unique()#all unique support count
fredic = {}
for i in range(len(su)):
    inset = list(frequent.loc[frequent.support ==su[i]]['itemsets'])
    fredic[su[i]] = inset
cl = []
for index, row in frequent.iterrows():
    isclose = True
    cli = row['itemsets']
    cls = row['support']
    checkset = fredic[cls]
    for i in checkset:
        if (cli!=i):
            if(frozenset.issubset(cli,i)):
                isclose = False
                break
    
    if(isclose):
        cl.append(row['itemsets'])
cl

C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


[frozenset({'Eggs'}),
 frozenset({'Donuts'}),
 frozenset({'Bananas'}),
 frozenset({'Carrots'}),
 frozenset({'Fish'}),
 frozenset({'Apples'}),
 frozenset({'Bananas', 'Eggs'}),
 frozenset({'Carrots', 'Eggs'}),
 frozenset({'Eggs', 'Fish'}),
 frozenset({'Bananas', 'Carrots', 'Eggs'}),
 frozenset({'Carrots', 'Eggs', 'Fish'}),
 frozenset({'Bananas', 'Eggs', 'Fish'}),
 frozenset({'Donuts', 'Eggs'}),
 frozenset({'Bananas', 'Donuts'}),
 frozenset({'Bananas', 'Carrots'}),
 frozenset({'Carrots', 'Fish'}),
 frozenset({'Bananas', 'Fish'}),
 frozenset({'Bananas', 'Carrots', 'Fish'}),
 frozenset({'Apples', 'Bananas'}),
 frozenset({'Apples', 'Fish'}),
 frozenset({'Apples', 'Donuts'}),
 frozenset({'Apples', 'Carrots', 'Fish'}),
 frozenset({'Apples', 'Bananas', 'Carrots', 'Fish'}),
 frozenset({'Apples', 'Bananas', 'Donuts'}),
 frozenset({'Apples', 'Donuts', 'Fish'}),
 frozenset({'Apples', 'Bananas', 'Eggs'})]

### 3I) How do the Closed Frequent Itemsets compare to the Maximal Frequent Itemsets?
- Maximial frequent itemsets is a subset of closed frequent itemsets

### 3J) When one might use the Closed Frequent vs the Maximal Frequent Itemsets?
- Goal: find the itemsets that closed but not maximal

***
# Section 4: Generate Rules (20%)
4A) How many possible rules are there for the data that exists ONLY in the dataset? For example, if there is no data set with items {A,B}, do not list any rules such as {A,B} -> C. 

4B) Calculate the number of rules possible if you had the itemset: {B, C, F}.

4C) List all the possible rules from the itemset shown in 4B.

4D) For the Itemset in 4B, using MinConf = 0.75 prune the rules using the anti-monotone property of rules. How many rules remain?

4E) List all the rules in 4D, and explain why.

4F) Explain why pruning rules might be advantageous for large data sets.
***

### 4A) Number of possible rules that exists ONLY in the dataset: 602

In [36]:
frequent_itemsets=mlx.frequent_patterns.apriori(ohe_df, min_support=0.001, use_colnames=True)
rules_ap = mlx.frequent_patterns.association_rules(frequent_itemsets, metric="support", min_threshold=0.001)
rules_ap

C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents                               consequents  antecedent support  \
0      (Apples)                                 (Bananas)                0.34   
1     (Bananas)                                  (Apples)                0.58   
2     (Carrots)                                  (Apples)                0.48   
3      (Apples)                                 (Carrots)                0.34   
4      (Apples)                                  (Donuts)                0.34   
..          ...                                       ...                 ...   
597   (Bananas)     (Carrots, Apples, Fish, Donuts, Eggs)                0.58   
598    (Apples)    (Carrots, Bananas, Fish, Donuts, Eggs)                0.34   
599      (Fish)  (Carrots, Bananas, Apples, Donuts, Eggs)                0.42   
600    (Donuts)    (Carrots, Bananas, Apples, Fish, Eggs)                0.40   
601      (Eggs)  (Carrots, Bananas, Apples, Fish, Donuts)                0.40   

     consequent support  support  confidence      lift  leverage  conviction  
0                  0.58     0.26    0.764706  1.318458    0.0628    1.785000  
1                  0.34     0.26    0.448276  1.318458    0.0628    1.196250  
2                  0.34     0.18    0.375000  1.102941    0.0168    1.056000  
3                  0.48     0.18    0.529412  1.102941    0.0168    1.105000  
4                  0.40     0.16    0.470588  1.176471    0.0240    1.133333  
..                  ...      ...         ...       ...       ...         ...  
597                0.02     0.02    0.034483  1.724138    0.0084    1.015000  
598                0.02     0.02    0.058824  2.941176    0.0132    1.041250  
599                0.02     0.02    0.047619  2.380952    0.0116    1.029000  
600                0.04     0.02    0.050000  1.250000    0.0040    1.010526  
601                0.02     0.02    0.050000  2.500000    0.0120    1.031579  

[602 rows x 9 columns]

### 4B) Calculate the number of rules possible if you had the itemset: {B, C, F}: 12

In [37]:
3**3-2**(3+1)+1

C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


12

### 4C) List all the possible rules from the itemset shown in 4B.

In [38]:
r1=rules_ap[(rules_ap['antecedents'] == {'Bananas'})&(rules_ap['consequents'] == {'Carrots'})]
r2=rules_ap[(rules_ap['antecedents'] == {'Bananas'})&(rules_ap['consequents'] == {'Fish'})]
r3=rules_ap[(rules_ap['antecedents'] == {'Bananas'})&(rules_ap['consequents'] == {'Carrots','Fish'})]
r4=rules_ap[(rules_ap['antecedents'] == {'Carrots'})&(rules_ap['consequents'] == {'Bananas'})]
r5=rules_ap[(rules_ap['antecedents'] == {'Carrots'})&(rules_ap['consequents'] == {'Fish'})]
r6=rules_ap[(rules_ap['antecedents'] == {'Carrots'})&(rules_ap['consequents'] == {'Bananas','Fish'})]
r7=rules_ap[(rules_ap['antecedents'] == {'Fish'})&(rules_ap['consequents'] == {'Bananas'})]
r8=rules_ap[(rules_ap['antecedents'] == {'Fish'})&(rules_ap['consequents'] == {'Carrots'})]
r9=rules_ap[(rules_ap['antecedents'] == {'Fish'})&(rules_ap['consequents'] == {'Bananas','Carrots'})]
r10=rules_ap[(rules_ap['antecedents'] == {'Bananas','Carrots'})&(rules_ap['consequents'] == {'Fish'})]
r11=rules_ap[(rules_ap['antecedents'] == {'Bananas','Fish'})&(rules_ap['consequents'] == {'Carrots'})]
r12=rules_ap[(rules_ap['antecedents'] == {'Carrots','Fish'})&(rules_ap['consequents'] == {'Bananas'})]

r=r1.append([r2,r3,r4,r5,r6,r7,r8,r9,r10,r11,r12])

r

C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents         consequents  antecedent support  \
11            (Bananas)           (Carrots)                0.58   
16            (Bananas)              (Fish)                0.58   
106           (Bananas)     (Carrots, Fish)                0.58   
10            (Carrots)           (Bananas)                0.48   
22            (Carrots)              (Fish)                0.48   
105           (Carrots)     (Bananas, Fish)                0.48   
17               (Fish)           (Bananas)                0.42   
23               (Fish)           (Carrots)                0.42   
107              (Fish)  (Carrots, Bananas)                0.42   
102  (Carrots, Bananas)              (Fish)                0.34   
104     (Bananas, Fish)           (Carrots)                0.26   
103     (Carrots, Fish)           (Bananas)                0.36   

     consequent support  support  confidence      lift  leverage  conviction  
11                 0.48     0.34    0.586207  1.221264    0.0616    1.256667  
16                 0.42     0.26    0.448276  1.067323    0.0164    1.051250  
106                0.36     0.24    0.413793  1.149425    0.0312    1.091765  
10                 0.58     0.34    0.708333  1.221264    0.0616    1.440000  
22                 0.42     0.36    0.750000  1.785714    0.1584    2.320000  
105                0.26     0.24    0.500000  1.923077    0.1152    1.480000  
17                 0.58     0.26    0.619048  1.067323    0.0164    1.102500  
23                 0.48     0.36    0.857143  1.785714    0.1584    3.640000  
107                0.34     0.24    0.571429  1.680672    0.0972    1.540000  
102                0.42     0.24    0.705882  1.680672    0.0972    1.972000  
104                0.48     0.24    0.923077  1.923077    0.1152    6.760000  
103                0.58     0.24    0.666667  1.149425    0.0312    1.260000

### 4D) 3

### 4E) List all the rules in 4D, and explain why.

In [39]:
r[r['confidence']>=0.75]

C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents consequents  antecedent support  consequent support  \
22         (Carrots)      (Fish)                0.48                0.42   
23            (Fish)   (Carrots)                0.42                0.48   
104  (Bananas, Fish)   (Carrots)                0.26                0.48   

     support  confidence      lift  leverage  conviction  
22      0.36    0.750000  1.785714    0.1584        2.32  
23      0.36    0.857143  1.785714    0.1584        3.64  
104     0.24    0.923077  1.923077    0.1152        6.76

- Reason: Confidence measures how often items in Y appear in transactions that contain X

- 4F) Explain why pruning rules might be advantageous for large data sets： Removing redundancy can reduce computational complexity.

***
# Section 5: Rule Evaluation (20%)
5A) For the rules in the dataset, find a rule (if any) that has a lift >1. Show the rule and the lift. 

5B) For the rules in the dataset, find a rule (if any) that has a lift <1. Show the rule and the lift. 

5C) How would the the lift change if in 5A) and 5B) if we added 1000 transactions to the dataset and each of these 1000 transactions had items that were not part of the original dataset (for example we started adding things like Teslas and yachts and other some such)? 

5d) What does the above tell you about lift? What other metrics might you consider?
***

### 5A) For the rules in the dataset, find a rule (if any) that has a lift >1. Show the rule and the lift: {Bananas}->{Apples} has the lift=1.318458

In [40]:
rules_5a=rules_ap[rules_ap['lift']>1]
rules_5a.reset_index(drop=True,inplace=True)
rules_5a

C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents                               consequents  antecedent support  \
0      (Apples)                                 (Bananas)                0.34   
1     (Bananas)                                  (Apples)                0.58   
2     (Carrots)                                  (Apples)                0.48   
3      (Apples)                                 (Carrots)                0.34   
4      (Apples)                                  (Donuts)                0.34   
..          ...                                       ...                 ...   
413   (Bananas)     (Carrots, Apples, Fish, Donuts, Eggs)                0.58   
414    (Apples)    (Carrots, Bananas, Fish, Donuts, Eggs)                0.34   
415      (Fish)  (Carrots, Bananas, Apples, Donuts, Eggs)                0.42   
416    (Donuts)    (Carrots, Bananas, Apples, Fish, Eggs)                0.40   
417      (Eggs)  (Carrots, Bananas, Apples, Fish, Donuts)                0.40   

     consequent support  support  confidence      lift  leverage  conviction  
0                  0.58     0.26    0.764706  1.318458    0.0628    1.785000  
1                  0.34     0.26    0.448276  1.318458    0.0628    1.196250  
2                  0.34     0.18    0.375000  1.102941    0.0168    1.056000  
3                  0.48     0.18    0.529412  1.102941    0.0168    1.105000  
4                  0.40     0.16    0.470588  1.176471    0.0240    1.133333  
..                  ...      ...         ...       ...       ...         ...  
413                0.02     0.02    0.034483  1.724138    0.0084    1.015000  
414                0.02     0.02    0.058824  2.941176    0.0132    1.041250  
415                0.02     0.02    0.047619  2.380952    0.0116    1.029000  
416                0.04     0.02    0.050000  1.250000    0.0040    1.010526  
417                0.02     0.02    0.050000  2.500000    0.0120    1.031579  

[418 rows x 9 columns]

### 5B) For the rules in the dataset, find a rule (if any) that has a lift <1. Show the rule and the lift: {Eggs}->{Apples} has the lift=0.441176

In [41]:
rules_5b=rules_ap[rules_ap['lift']<1]
rules_5b.reset_index(drop=True,inplace=True)
rules_5b

C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents                       consequents  \
0                            (Apples)                            (Eggs)   
1                              (Eggs)                          (Apples)   
2                           (Bananas)                          (Donuts)   
3                            (Donuts)                         (Bananas)   
4                           (Bananas)                            (Eggs)   
..                                ...                               ...   
179                    (Donuts, Eggs)           (Carrots, Apples, Fish)   
180          (Carrots, Bananas, Fish)                    (Donuts, Eggs)   
181                    (Donuts, Eggs)          (Carrots, Bananas, Fish)   
182  (Carrots, Bananas, Fish, Apples)                    (Donuts, Eggs)   
183                    (Donuts, Eggs)  (Carrots, Bananas, Fish, Apples)   

     antecedent support  consequent support  support  confidence      lift  \
0                  0.34                0.40     0.06    0.176471  0.441176   
1                  0.40                0.34     0.06    0.150000  0.441176   
2                  0.58                0.40     0.10    0.172414  0.431034   
3                  0.40                0.58     0.10    0.250000  0.431034   
4                  0.58                0.40     0.22    0.379310  0.948276   
..                  ...                 ...      ...         ...       ...   
179                0.16                0.18     0.02    0.125000  0.694444   
180                0.24                0.16     0.02    0.083333  0.520833   
181                0.16                0.24     0.02    0.125000  0.520833   
182                0.16                0.16     0.02    0.125000  0.781250   
183                0.16                0.16     0.02    0.125000  0.781250   

     leverage  conviction  
0     -0.0760    0.728571  
1     -0.0760    0.776471  
2     -0.1320    0.725000  
3     -0.1320    0.560000  
4     -0.0120    0.966667  
..        ...         ...  
179   -0.0088    0.937143  
180   -0.0184    0.916364  
181   -0.0184    0.868571  
182   -0.0056    0.960000  
183   -0.0056    0.960000  

[184 rows x 9 columns]

### 5C) How would the the lift change if in 5A) and 5B) if we added 1000 transactions to the dataset and each of these 1000 transactions had items that were not part of the original dataset (for example we started adding things like Teslas and yachts and other some such)?

In [42]:
rules_ap[rules_ap['antecedents']=={'Bananas','Apples'}]

C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents                    consequents  antecedent support  \
32   (Apples, Bananas)                      (Carrots)                0.26   
37   (Apples, Bananas)                       (Donuts)                0.26   
43   (Apples, Bananas)                         (Eggs)                0.26   
49   (Apples, Bananas)                         (Fish)                0.26   
158  (Apples, Bananas)              (Carrots, Donuts)                0.26   
172  (Apples, Bananas)                (Carrots, Eggs)                0.26   
186  (Apples, Bananas)                (Carrots, Fish)                0.26   
198  (Apples, Bananas)                 (Donuts, Eggs)                0.26   
212  (Apples, Bananas)                 (Fish, Donuts)                0.26   
226  (Apples, Bananas)                   (Fish, Eggs)                0.26   
379  (Bananas, Apples)        (Carrots, Donuts, Eggs)                0.26   
409  (Bananas, Apples)        (Carrots, Fish, Donuts)                0.26   
439  (Bananas, Apples)          (Carrots, Fish, Eggs)                0.26   
465  (Bananas, Apples)           (Fish, Donuts, Eggs)                0.26   
586  (Bananas, Apples)  (Carrots, Fish, Donuts, Eggs)                0.26   

     consequent support  support  confidence      lift  leverage  conviction  
32                 0.48     0.16    0.615385  1.282051    0.0352    1.352000  
37                 0.40     0.08    0.307692  0.769231   -0.0240    0.866667  
43                 0.40     0.06    0.230769  0.576923   -0.0440    0.780000  
49                 0.42     0.16    0.615385  1.465201    0.0508    1.508000  
158                0.04     0.02    0.076923  1.923077    0.0096    1.040000  
172                0.16     0.04    0.153846  0.961538   -0.0016    0.992727  
186                0.36     0.16    0.615385  1.709402    0.0664    1.664000  
198                0.16     0.02    0.076923  0.480769   -0.0216    0.910000  
212                0.08     0.02    0.076923  0.961538   -0.0008    0.996667  
226                0.10     0.04    0.153846  1.538462    0.0140    1.063636  
379                0.02     0.02    0.076923  3.846154    0.0148    1.061667  
409                0.04     0.02    0.076923  1.923077    0.0096    1.040000  
439                0.08     0.04    0.153846  1.923077    0.0192    1.087273  
465                0.02     0.02    0.076923  3.846154    0.0148    1.061667  
586                0.02     0.02    0.076923  3.846154    0.0148    1.061667

### New lift for {Bananas}->{Apples} is: ((50 × 0.26)/1050)/((50 × 0.34)/1050) ×((50 × 0.58)/1050)=27.688


In [43]:
rules_ap[rules_ap['antecedents']=={'Eggs','Apples'}]

C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents                       consequents  antecedent support  \
42   (Apples, Eggs)                         (Bananas)                0.06   
62   (Apples, Eggs)                         (Carrots)                0.06   
73   (Apples, Eggs)                          (Donuts)                0.06   
85   (Apples, Eggs)                            (Fish)                0.06   
171  (Apples, Eggs)                (Carrots, Bananas)                0.06   
197  (Apples, Eggs)                 (Bananas, Donuts)                0.06   
225  (Apples, Eggs)                   (Bananas, Fish)                0.06   
242  (Apples, Eggs)                 (Carrots, Donuts)                0.06   
270  (Apples, Eggs)                   (Carrots, Fish)                0.06   
282  (Apples, Eggs)                    (Fish, Donuts)                0.06   
383  (Apples, Eggs)        (Carrots, Bananas, Donuts)                0.06   
443  (Apples, Eggs)          (Carrots, Bananas, Fish)                0.06   
471  (Apples, Eggs)           (Bananas, Fish, Donuts)                0.06   
501  (Apples, Eggs)           (Carrots, Fish, Donuts)                0.06   
592  (Apples, Eggs)  (Carrots, Bananas, Fish, Donuts)                0.06   

     consequent support  support  confidence       lift  leverage  conviction  
42                 0.58     0.06    1.000000   1.724138    0.0252         inf  
62                 0.48     0.04    0.666667   1.388889    0.0112        1.56  
73                 0.40     0.02    0.333333   0.833333   -0.0040        0.90  
85                 0.42     0.04    0.666667   1.587302    0.0148        1.74  
171                0.34     0.04    0.666667   1.960784    0.0196        1.98  
197                0.10     0.02    0.333333   3.333333    0.0140        1.35  
225                0.26     0.04    0.666667   2.564103    0.0244        2.22  
242                0.04     0.02    0.333333   8.333333    0.0176        1.44  
270                0.36     0.04    0.666667   1.851852    0.0184        1.92  
282                0.08     0.02    0.333333   4.166667    0.0152        1.38  
383                0.02     0.02    0.333333  16.666667    0.0188        1.47  
443                0.24     0.04    0.666667   2.777778    0.0256        2.28  
471                0.02     0.02    0.333333  16.666667    0.0188        1.47  
501                0.04     0.02    0.333333   8.333333    0.0176        1.44  
592                0.02     0.02    0.333333  16.666667    0.0188        1.47

### New lift for {Eggs}->{Apples} is: ((50 × 0.06)/1050)/((50 × 0.40)/1050) ×((50 ×0.34)/1050)=9.26


### 5d) What does the above tell you about lift: When the size of transactions increases, the lift would increase for both lift >1 and <1.


### What other metrics might you consider: Symmetric measures: support, collective strength, cosine, Jaccard, etc. Asymmetric measures: confidence, conviction, Laplace, J-measure, etc

***
# Section 6: Other Support Metrics (bonus) (20%)
6A) Consider our favorite itemset of {B, C, F}. Compute the cross-support r(x) with explanation.

6B) Compute the hconf of the itemset in 6A) with explanation.

6C) Compute the hyperclique (if the threshold is 50%) for the items in part 6A).
***

### 6A) Consider our favorite itemset of {B, C, F}. Compute the cross-support r(x) with explanation: r(X)=min{s(B),s(C),s(F)}/max{s(B),s(C),s(F)} = 0.42/0.58= 0.724

### 6B) Compute the hconf of the itemset in 6A) with explanation: hconf(X) = min( conf(X1→ X2) )=0.413793

In [44]:
r

C:\Users\xiang\anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents         consequents  antecedent support  \
11            (Bananas)           (Carrots)                0.58   
16            (Bananas)              (Fish)                0.58   
106           (Bananas)     (Carrots, Fish)                0.58   
10            (Carrots)           (Bananas)                0.48   
22            (Carrots)              (Fish)                0.48   
105           (Carrots)     (Bananas, Fish)                0.48   
17               (Fish)           (Bananas)                0.42   
23               (Fish)           (Carrots)                0.42   
107              (Fish)  (Carrots, Bananas)                0.42   
102  (Carrots, Bananas)              (Fish)                0.34   
104     (Bananas, Fish)           (Carrots)                0.26   
103     (Carrots, Fish)           (Bananas)                0.36   

     consequent support  support  confidence      lift  leverage  conviction  
11                 0.48     0.34    0.586207  1.221264    0.0616    1.256667  
16                 0.42     0.26    0.448276  1.067323    0.0164    1.051250  
106                0.36     0.24    0.413793  1.149425    0.0312    1.091765  
10                 0.58     0.34    0.708333  1.221264    0.0616    1.440000  
22                 0.42     0.36    0.750000  1.785714    0.1584    2.320000  
105                0.26     0.24    0.500000  1.923077    0.1152    1.480000  
17                 0.58     0.26    0.619048  1.067323    0.0164    1.102500  
23                 0.48     0.36    0.857143  1.785714    0.1584    3.640000  
107                0.34     0.24    0.571429  1.680672    0.0972    1.540000  
102                0.42     0.24    0.705882  1.680672    0.0972    1.972000  
104                0.48     0.24    0.923077  1.923077    0.1152    6.760000  
103                0.58     0.24    0.666667  1.149425    0.0312    1.260000

### 6C) Compute the hyperclique (if the threshold is 50%) for the items in part 6A): A hyperclique X is maximal if hconf(X)<=hc and none of its immediate supersets, Y, have hconf(Y)<=hc. So {Bananas,Carrots,Fish} and {Bananas,Fish} are hypercliques.